packages/transformers/webextension/src/WebExtensionTransformer.js

```diff
@@ -266,91 +266,81 @@ async function collectDependencies(
-          program.background = {};
-        }
-        if (!program.background.scripts) {
-          program.background.scripts = [];
-        }
-        if (program.background.scripts.length == 0) {
-          program.background.scripts.push(
-            asset.addURLDependency('./runtime/default-bg.js', {
-              resolveFrom: __filename,
-            }),
-          );
-        }
-        asset.meta.webextBGInsert = program.background.scripts[0];
+    if (needRuntimeBG && !program.background?.page) {
+      if (!program.background) {
+        program.background = {};
+      }
+      if (!program.background.scripts) {
+        program.background.scripts = [];
+      }
+      if (program.background.scripts.length == 0) {
+        program.background.scripts.push(
+          asset.addURLDependency('./runtime/default-bg.js', {
+            resolveFrom: __filename,
+          }),
+        );
       }
+      asset.meta.webextBGInsert = program.background.scripts[0];
     }
-  } else {
-    if (program.background?.service_worker) {
-      program.background.service_worker = asset.addURLDependency(
-        program.background.service_worker,
-        {
-          bundleBehavior: 'isolated',
-          loc: {
-            filePath,
-            ...getJSONSourceLocation(
-              ptrs['/background/service_worker'],
-              'value',
-            ),
-          },
-          env: {
-            context: 'service-worker',
-            sourceType:
-              program.background.type == 'module' ? 'module' : 'script',
-          },
+  }
+  if (program.background?.service_worker) {
+    program.background.service_worker = asset.addURLDependency(
+      program.background.service_worker,
+      {
+        bundleBehavior: 'isolated',
+        loc: {
+          filePath,
+          ...getJSONSourceLocation(ptrs['/background/service_worker'], 'value'),
         },
-      );
-    }
-    if (hot) {
-      // Enable eval HMR for sandbox,
-      const csp = program.content_security_policy || {};
-      csp.extension_pages = cspPatchHMR(
-        csp.extension_pages,
-        `http://${hmrOptions?.host || 'localhost'}`,
-      );
-      // Sandbox allows eval by default
-      if (csp.sandbox) csp.sandbox = cspPatchHMR(csp.sandbox);
-      program.content_security_policy = csp;
-      if (needRuntimeBG) {
-        if (!program.background) {
-          program.background = {};
-        }
-        if (!program.background.service_worker) {
-          program.background.service_worker = asset.addURLDependency(
-            './runtime/default-bg.js',
-            {
-              resolveFrom: __filename,
-              env: {context: 'service-worker'},
-            },
-          );
-        }
-        asset.meta.webextBGInsert = program.background.service_worker;
+        env: {
+          context: 'service-worker',
+          sourceType: program.background.type == 'module' ? 'module' : 'script',
+        },
+      },
+    );
+  }
+  if (hot) {
```

Comment:
There are now two separate blocks of HMR handling code. This is going to cause issues with HMR: Chrome requires different handling of v2 and v3 to do HMR properly. I'm not sure which version is needed for Firefox's version of Manifest V3. Could you try to investigate?

Reply:
One way to fix the double-handling of `hot` would be to keep the `isMV2` check but copy the `background.scripts` code into the `else` block as well. But you'd still need to investigate if this HMR code actually works in Firefox.

Reply:
Thanks for review! I have pushed the commit with the fix. https://github.com/parcel-bundler/parcel/pull/8906/commits/1fbbe8bd6f39b94ca2b4ebe2be2964231a800621

I investigated whether HMR works in Firefox and having trouble getting HMR to work in Firefox.

In Manifest V3, [the script-src directives may only have 'self', 'none' and 'wasm-unsafe-eval'. ](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/content_security_policy#manifest_v3_syntax)



Overwrite at the following value in the `dist/manifest.json` file after the build, It will work HMR when saving the files specified in the `background.scripts`.

```
  "content_security_policy": {
    "extension_pages": "script-src 'self';"
  },
```

But, occur the following errors when saving the files specified in the content scripts

<img width="968" alt="Screen Shot 2023-05-10 at 2 39 55" src="https://github.com/parcel-bundler/parcel/assets/17245737/424cf57f-fe0f-4098-aa40-d7f1dcf75fb7">

It's possible that the reason is the stricter CSP constraints in Manifest V3.

> In Manifest V3, all CSP sources that refer to external or non-static content are forbidden. The only permitted values are 'none', 'self', and 'wasm-unsafe-eval'. In Manifest V2, a source for a script directive is considered secure if it meets these criteria:
https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/content_security_policy